iree/compiler/Dialect/IREE/IR/IREEInterfaces.td

```diff
@@ -0,0 +1,137 @@
+    way. Results are still SSA values distinct from the operands and the tie is
+    strictly a relationship relevant to transformations and not something that
+    modifies IR definitions.
+
+    Example:
+      An operation on tensors that wants to indicate that the storage for a
+      result should alias the storage for an operand, performing an "in-place"
+      operation. Since tensors are still used there is no hard requirement that
+      uses of the result SSA value alias the operand; a copy may still be
+      introduced.
+
+      See: flow.dispatch.workgroups
+
+    Example:
+      An operation on buffers that wants to encode activity on the buffer in IR
+      (such as a barrier, a transfer operation, etc) such that the SSA use-def
+      chain is representing the state of the buffer at various points in time
+      but that the underlying buffers are all tied together.
+
+      See: hal.stream.barrier
+
+    The default implementations use an attribute on the op to store the
+    relationship:
+      `OptionalAttr<IREE_TiedOpStorageAttr>:$tied_operands`
+  }];
+
+  let methods = [
+    InterfaceMethod<[{
+      Returns the set of operands that results may be tied to.
+      }],
+      "std::pair<unsigned, unsigned>", "getTiedOperandsIndexAndLength", (ins),
+      [{}],
+      [{
+        return {0, $_op.getNumOperands()};
+      }]
+    >,
+    InterfaceMethod<[{
+      Walks the SSA use-def chain to find the first defined value reachable
+      from the given value by traversing tied ops. Note that the returned
+      value may be a block argument and have no defining op.
+      }],
+      "Value", "getTiedResult",
+      (ins "unsigned":$resultIndex),
+      [{}],
+      [{
+        return IREE::TiedOpInterface::findTiedBaseValue($_op.getResult(resultIndex));
+      }]
+    >,
+    InterfaceMethod<[{
+      Returns the operand tied to the given result of the op or nullptr if none.
+      }],
+      "Value", "getTiedResultOperand",
+      (ins "Value":$result),
+      [{}],
+      [{
+        auto resultIndex = std::distance($_op->getResults().begin(),
+                                         llvm::find($_op->getResults(), result));
+        auto operandIndex = IREE::detail::getTiedResultOperandIndex($_op, resultIndex);
+        return operandIndex.hasValue() ?
+            $_op.getOperand(operandIndex.getValue()) :
+            nullptr;
+      }]
+    >,
+    InterfaceMethod<[{
+      Returns the operand index tied to the given result index, if any.
+      }],
+      "::llvm::Optional<unsigned>", "getTiedResultOperandIndex",
+      (ins "unsigned":$resultIndex),
+      [{}],
+      [{
+        return IREE::detail::getTiedResultOperandIndex($_op, resultIndex);
+      }]
+    >,
+    InterfaceMethod<[{
+      Sets the operand index tied to the given result index, if any.
+      }],
+      "void", "setTiedResultOperandIndex",
+      (ins "unsigned":$resultIndex, "::llvm::Optional<unsigned>":$operandIndex),
```

Comment:
Not sure I follow - the optional is because you can either set an operand index that the result is tied to _or_ clear it and indicate that the result is not tied to anything. What in ODS are you referring to?

Reply:
Ok, maybe I am mistaken on what this is intended on. I thought it was an "optional" argument with a default value.

Reply:
ah no - just a way to clear things that's paired with getTiedResultOperandIndex that also returns an Optional<unsigned> - that way you can do `opB.setTiedResultOperandIndex(0, opA.getTiedResultOperandIndex(0))`